### PR TITLE
Workaround broken DocumentFragment in older versions of IE 11

### PR DESCRIPTION
**Pull request by dfreedm**
In these old versions of IE 11, modifying a DocumentFragment while a
MutationObserver is active could leave the childNodes of the
DocumentFragment with either a `null` parentNode, or a parentNode
pointing to a new DocumentFragment wrapper that is not identical to the
current one, but shares prototypical values like `childNodes`.

This can lead to instances where `frag.firstChild.parentNode !== frag`,
but `frag.firstChild.parentNode.childNodes === frag.childNodes`.
Eventually, this new wrapper is GC'd and the parentNode will be null,
causing all sorts of SUPER FUN PROBLEMS.

### CODE REVIEW

**Comment by ruud** [comment]
Would be great to have this in

**Comment by web-padawan** [comment]
@azakus I was wondering if this should be backported to 1.x or not?
Your description looks somewhat similar to [polymer/4455](https://github.com/Polymer/polymer/issues/4455)

**Comment by dfreedm** [comment]
@web-padawan yeah shouldn't be too hard to port back to webcomponents v0

File: template.js [hunk 1]
```diff
@@ -13,26 +13,109 @@
...
-          f.appendChild(n);
-          return f;
+
+      needsDocFrag = true;
+
+      var origCloneNode = Node.prototype.cloneNode;
+      Node.prototype.cloneNode = function cloneNode(deep) {
+        if (this instanceof DocumentFragment) {
+          var newFrag = origCloneNode.call(this, deep);
+          if (this instanceof DocumentFragment) {
```
Comment (sorvell) [inline template.js]: dup, remove

File: template.js [hunk 2]
```diff
@@ -229,55 +329,55 @@
...
 
-    PolyfilledHTMLTemplateElement.prototype.cloneNode = function(deep) {
-      return PolyfilledHTMLTemplateElement._cloneNode(this, deep);
-    }
-
     // Given a source and cloned subtree, find <template>'s in the cloned
     // subtree and replace them with cloned <template>'s from source.
     // We must do this because only the source templates have proper .content.
-    PolyfilledHTMLTemplateElement.fixClonedDom = function(clone, source) {
+    PolyfilledHTMLTemplateElement._fixClonedDom = function _fixClonedDom(clone, source) {
```
Comment (sorvell) [inline template.js]: Let's just not expose this method?

File: template.js [hunk 3]
```diff
@@ -287,23 +387,18 @@
...
+        var dom = capturedImportNode.call(this, element, deep);
         if (deep) {
-          PolyfilledHTMLTemplateElement.fixClonedDom(dom, element);
+          PolyfilledHTMLTemplateElement._fixClonedDom(dom, element);
         }
         return dom;
       }
     };
-
-    if (needsCloning) {
```
Comment (sorvell) [inline template.js:302]: Why is this removed?

**Review by sorvell — CHANGES_REQUESTED** [review]

File: template.js (hunk 3 above)
Comment (dfreedm) [reply template.js:302]: Not needed since I'm overriding `Node.prototype.cloneNode` entirely.

File: template.js (hunk 2 above)
Comment (dfreedm) [reply template.js]: Ok
